### PR TITLE
[service.subtitles.supersubtitles] 0.0.23

### PR DESCRIPTION
**Pull request by fape**
### Description
- archive handling changes for kodi leia
### Checklist:
- [X] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-scripts/blob/master/CONTRIBUTING.md) document
- [X] Each add-on submission should be a single commit with using the following style: [script.foo.bar] 1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practise but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-scripts/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

**Comment by fape** [comment]
> This add-on isn't Python 3 compatible. As mentioned at https://kodi.tv/article/attention-addon-developers-migration-python-3 only Python 3 compatible add-ons are accepted into the leia branch.

Could please help, what code piece code/part is not python 3 compatible? How can I test my addon with python3? Thanks for your advice.

**Comment by Rechi** [comment]
This is the error I got while running this add-on. There might be more errors, which will show up once this one is fixed.
```
ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
 - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
Error Type: <class 'ModuleNotFoundError'>
Error Contents: No module named 'urllib2'
Traceback (most recent call last):
  File ".../addons/service.subtitles.supersubtitles/service.py", line 23, in <module>
    import urllib2
ModuleNotFoundError: No module named 'urllib2'
-->End of Python script error report<--
```

**Comment by fape** [comment]
> This is the error I got while running this add-on. There might be more errors, which will show up once this one is fixed.
> 
> ```
> ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
>  - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
> Error Type: <class 'ModuleNotFoundError'>
> Error Contents: No module named 'urllib2'
> Traceback (most recent call last):
>   File ".../addons/service.subtitles.supersubtitles/service.py", line 23, in <module>
>     import urllib2
> ModuleNotFoundError: No module named 'urllib2'
> -->End of Python script error report<--
> ```

How did you run in with python3?
I testes with latest stable kodi 18.3 (leia) with win64 and osmc build without anyproblem.
Are you using nightly for koid 19?

**Comment by fape** [comment]
I found one build with python3. But do not get is why I have to create script which is supported by python2 and python3 also. I have to add lot of boilerplate version check if statement, mainly because of the string handling changes... But my script will run on python2 in kodi 18. 

I'm going to lett my users to why is rar handling broken (because of kodi changes).

Please close this PR or I'll do. I do not have time and motivation to create boilerplate code.

**Comment by enen92** [comment]
@fape python2 is EOL in 2020, so it will break for the upcoming Kodi Matrix release which will ship python 3 only. If you don't want to make it accessible via the official repository it's entirely up to you but in case you want to do so you have to keep the addon both py2 and py3 compatible for it to be in the Leia branch. Hopefully when the Matrix branch is created addons in that branch only need to be py3 compatible.

**Comment by fape** [comment]
> @fape python2 is EOL in 2020, so it will break for the upcoming Kodi Matrix release which will ship python 3 only. If you don't want to make it accessible via the official repository it's entirely up to you but in case you want to do so you have to keep the addon both py2 and py3 compatible for it to be in the Leia branch. Hopefully when the Matrix branch is created addons in that branch only need to be py3 compatible.

I understand. But I do not agree with that. Does not make sense to create and maintain addon which supports python2 and python3 at same time...
Why should we create lots of boilerplate code to support leila with python3. Will be any kodi leia release with python3? No so why we have to support that?

**Comment by enen92** [comment]
@fape because Matrix will have access to all repos since gotham. And alpha users do exist.

**Comment by fape** [comment]
> @fape because Matrix will have access to all repos since gotham. And alpha users do exist.

What kind of alpha users? The [Move from python 2 to python 3 ](https://github.com/xbmc/xbmc/pull/16116) is still open and is not in mergeable state... So now maybe some harcore users is using python3...
When matrix nightly contains python3 I'm going to refactor my script to use ONLY python3 and create PR to matrix branch.

**Comment by fape** [comment]
@Rechi is it OK for you now?

File: service.subtitles.supersubtitles/addon.xml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<addon id="service.subtitles.supersubtitles"
+       name="Super Subtitles"
+       version="0.0.23"
+       provider-name="fape">
+  <requires>
+    <import addon="xbmc.python" version="2.26.0"/>
+    <import addon="xbmc.addon" version="17.9.910"/>
```
Comment (enen92) [inline service.subtitles.supersubtitles/addon.xml]: please remove the `xbmc.addon` dependency, not needed

File: service.subtitles.supersubtitles/addon.xml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+             library="service.py" />
+  <extension point="xbmc.addon.metadata">
+    <summary lang="en_GB">Super Subtitles</summary>
+    <description lang="en_GB">Search and download subtitles from Super Subtitles (feliratok.info)</description>
+    <platform>all</platform>
+    <license>GNU GENERAL PUBLIC LICENSE. Version 2, June 1991</license>
+    <forum>https://forum.kodi.tv/showthread.php?tid=188591</forum>
+    <website>https://www.feliratok.info/index.php</website>
+    <email>fape88 at gmail dot com</email>
+    <language></language>
```
Comment (enen92) [inline service.subtitles.supersubtitles/addon.xml]: please remove the empty language tag

File: service.subtitles.supersubtitles/addon.xml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    <import addon="script.module.kodi-six" version="0.1.0"/>
+    <import addon="vfs.libarchive" version="1.0.5"/>
+  </requires>
+  <extension point="xbmc.subtitle.module"
+             library="service.py" />
+  <extension point="xbmc.addon.metadata">
+    <summary lang="en_GB">Super Subtitles</summary>
+    <description lang="en_GB">Search and download subtitles from Super Subtitles (feliratok.info)</description>
+    <platform>all</platform>
+    <license>GNU GENERAL PUBLIC LICENSE. Version 2, June 1991</license>
```
Comment (enen92) [inline service.subtitles.supersubtitles/addon.xml]: It'd be nice to have an spdx identifier instead: `GPL-2.0`

File: service.subtitles.supersubtitles/changelog.txt [hunk 4]
```diff
@@ -0,0 +1,93 @@
+0.0.23
```
Comment (enen92) [inline service.subtitles.supersubtitles/changelog.txt]: The changelog is no longer parsed by kodi. You can still keep it in the submission as a global changelog, but Kodi now has a field in addon.xml dedicated to this. You can add:

```xml
<news>
    0.0.23:
      - archive handling changes
      - leia addon related changes
      - inital python3 support
</news>
```
To the extension.metadata of your addon.xml. This news field is dedicated to the changelog of the last version only.

File: service.subtitles.supersubtitles/addon.xml (hunk 2 above)
Comment (fape) [reply service.subtitles.supersubtitles/addon.xml]: There are 8 addons in leai branch with empty language tag.

File: service.subtitles.supersubtitles/addon.xml (hunk 3 above)
Comment (fape) [reply service.subtitles.supersubtitles/addon.xml]: there are 10 addons in leai branch such a long license description

File: service.subtitles.supersubtitles/changelog.txt (hunk 4 above)
Comment (fape) [reply service.subtitles.supersubtitles/changelog.txt]: I know that.

**Comment by fape** [comment]
@enen92  Any real problem?

File: service.subtitles.supersubtitles/addon.xml (hunk 2 above)
Comment (fape) [reply service.subtitles.supersubtitles/addon.xml]: [What about offical documentation](https://kodi.wiki/view/Addon.xml#.3Clanguage.3E):

> When there is no specific language provided in your content, leave it blank.

File: service.subtitles.supersubtitles/addon.xml (hunk 3 above)
Comment (fape) [reply service.subtitles.supersubtitles/addon.xml]: Again: [official documentation](https://kodi.wiki/view/Addon.xml#.3Clicense.3E)

> <license>GNU GENERAL PUBLIC LICENSE. Version 2, June 1991</license>

Comment (enen92) [reply service.subtitles.supersubtitles/addon.xml]: **it would be nice** != Do it

File: service.subtitles.supersubtitles/addon.xml (hunk 2 above)
Comment (enen92) [reply service.subtitles.supersubtitles/addon.xml]: Subtitle addons do not provide content.

**Comment by enen92** [comment]
Yes, you. I'm out. 
For this PR to go in it has to be reviewed by someone else.

**Comment by fape** [comment]
@enen92: I do sincerely apologize for my tone in the previous comments, I honestly did not intend to sound as harsh as it come out.

File: service.subtitles.supersubtitles/changelog.txt (hunk 4 above)
Comment (basrieter) [reply service.subtitles.supersubtitles/changelog.txt]: The changelog.txt is no longer used by Kodi and should not be included in an add-on. If you want user to see it, has to be put in the addon.xml `<news>`. If you don't think a users need this, then don't submit the changelog.txt.

File: service.subtitles.supersubtitles/service.py [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+if py2:
+    from urllib import unquote, unquote_plus, urlencode, quote_plus
+    from urllib2 import urlopen, Request, HTTPError, URLError
+else :
+    from urllib.parse import unquote, unquote_plus, urlencode, quote_plus
+    from urllib.request import urlopen, Request
+    from urllib.error import HTTPError, URLError
+
+__addon__ = xbmcaddon.Addon()
+#__author__ = __addon__.getAddonInfo('author')
```
Comment (basrieter) [inline service.subtitles.supersubtitles/service.py]: I would _suggest_ to not include commented code lines in a release version. You can always look at your own source control history to see what was there before.

File: service.subtitles.supersubtitles/service.py [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+    downloaded = download_file(item)
+
+    if is_archive(downloaded):
+        debuglog('%s downloaded file is an archive' % downloaded)
+        archive = 'archive://%s' % quote_plus(downloaded)
+        subtitle = recursive_search(archive)
+
+        if not subtitle:
+            debuglog("No subtitle found by search. Open dialog from %s" % archive)
+            #dialog = xbmcgui.Dialog()
```
Comment (basrieter) [inline service.subtitles.supersubtitles/service.py]: Same here: the commented code is not useful at all.

File: service.subtitles.supersubtitles/service.py [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+def lang_hun2eng(hunlang):
+    return LANGUAGES[py2_encode(hunlang, "utf-8").lower()]
+
+
+def log(msg, level):
+    xbmc.log(py2_encode((u"### [%s] - %s" % (__scriptid__, msg)), 'utf-8'), level=level)
+
+
+def infolog(msg):
+    log(msg, xbmc.LOGNOTICE)
```
Comment (basrieter) [inline service.subtitles.supersubtitles/service.py]: Not allowed. From the rules: _addons should use the xbmc.log() method to write to the logfile and use the debug logging level only. print statements should not be used._

File: service.subtitles.supersubtitles/service.py [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+
+
+def errorlog(msg):
+    log(msg, xbmc.LOGERROR)
```
Comment (basrieter) [inline service.subtitles.supersubtitles/service.py]: Not allowed. From the rules: _addons should use the xbmc.log() method to write to the logfile and use the debug logging level only. print statements should not be used._

File: service.subtitles.supersubtitles/service.py [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+    except URLError as e:
+        errorlog("URL Error %s, %s" % (e.reason, url))
+    except Exception as e:
+        errorlog("Unexpected exception: %s" % e.message)
+
+    return None
+
+
+def query_data(params):
+    response = send_request(params)
```
Comment (basrieter) [inline service.subtitles.supersubtitles/service.py]: You should use `with send_request(params) as reponse:` to make sure all resources are release after exiting the block. If not, you will be leaking memory which is bad for a service.

File: service.subtitles.supersubtitles/service.py [hunk 10]
```diff
@@ -0,0 +1,577 @@
...
+
+def search(item):
+    debuglog(item)
+    subtitles_list = search_subtitles(item)
+
+    if subtitles_list:
+        index = 0
+        for it in subtitles_list:
+            index += 1
+            #label="%s | %s | %s"%(it['name'], it['filename'], it['uploader'])
```
Comment (basrieter) [inline service.subtitles.supersubtitles/service.py]: Again: commented code

File: service.subtitles.supersubtitles/service.py [hunk 11]
```diff
@@ -0,0 +1,577 @@
...
+                return True
+    return False
+
+
+def download_file(item):
+    filename = py2_decode((unquote_plus(py2_decode(item['filename'], "utf-8")).replace(' ', '_')), 'utf-8')
+    localfile = os.path.join(__temp__, filename)
+    qparams = {'action': 'letolt', 'felirat': item['id']}
+
+    response = send_request(qparams)
```
Comment (basrieter) [inline service.subtitles.supersubtitles/service.py]: Same story with the `with send_request(qparams) as reponse;` to make sure resources are released

File: service.subtitles.supersubtitles/service.py [hunk 12]
```diff
@@ -0,0 +1,577 @@
...
+            return json.loads(response.read(), encoding='utf-8')
+        except ValueError as e:
+            errorlog("Json Decode Error: %s" % e.message)
+        except Exception as e:
+            errorlog("Unexpected exception: %s" % e.message)
+    return None
+
+
+def notification(id):
+    xbmc.executebuiltin(u'Notification(%s,%s,%s,%s)' % (
```
Comment (Rechi) [inline service.subtitles.supersubtitles/service.py]: use `xbmcgui.Dialog().notification(heading, message[, icon, time, sound])` (https://codedocs.xyz/xbmc/xbmc/group__python___dialog.html#ga27157f98dddb21b44cc6456137977aa2)

**Comment by fape** [comment]
> I also wonder were the `logo.png` is used? I could not see where. If it is not used, please remove it.

logo.png is used by [GUIDialogSubtitles.cpp](https://github.com/xbmc/xbmc/blob/770c691bfeea88d561b39d2e38348ac434db3adc/xbmc/video/dialogs/GUIDialogSubtitles.cpp#L279)

File: service.subtitles.supersubtitles/service.py (hunk 9 above)
Comment (fape) [reply service.subtitles.supersubtitles/service.py]: unfortunately urllib2 (in python2) is not supported by `with `statement. I'm going to find other solution.

Comment (basrieter) [reply service.subtitles.supersubtitles/service.py]: Ok, they you can use a `try: finally:` block with a `close()` in the finally

Comment (fape) [reply service.subtitles.supersubtitles/service.py]: I managed to use contextlib.closing:
`with contextlib.closing(send_request(params)) as response:`

Comment (basrieter) [reply service.subtitles.supersubtitles/service.py]: Indeed, same result.

**Pull request merged**